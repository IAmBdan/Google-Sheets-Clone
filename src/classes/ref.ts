class Ref {
  column: string;
  row: number;

  constructor(ref: string);
  constructor(column: string, row: number);
  constructor(refOrColumn: string, row?: number) {
    if (typeof row === "undefined") {
      // If row is undefined, assume refOrColumn is the reference string
      const match = /^\$([A-Za-z]+)(\d+)$/.exec(refOrColumn.toUpperCase());
      if (!match?.[1] || !match[2]) {
        throw new Error(`Invalid ref: ${refOrColumn}`);
      }
      this.column = match[1];
      this.row = parseInt(match[2], 10);
    } else {
      // If row is defined, assume refOrColumn is the column string
      this.column = refOrColumn;
      this.row = row;
    }
  }

  toString(): string {
    return `$${this.column}${this.row}`;
  }

  isEqual(ref: Ref): boolean {
    return this.column === ref.column && this.row === ref.row;
  }

  getColumn(): string {
    return this.column;
  }

  getRow(): number {
    return this.row;
  }

  //converts the column to a number (A=1, B=2, AA=27, AB = 28, etc)
  getColumnIndex(): number {
    return (
      this.column
        .split("")
        .reduce((acc, char) => acc * 26 + char.charCodeAt(0) - 64, 0) - 1
    );
  }

  setColumn(column: string): void {
    this.column = column;
  }

  setRow(row: number): void {
    this.row = row;
  }
}

export default Ref;
