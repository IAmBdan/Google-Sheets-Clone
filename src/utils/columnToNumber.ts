

export function columnToNumber(label: string): number {
    label = label.toUpperCase();
    let columnNumber = 0;
    for (let i = 0; i < label.length; i++) {
        columnNumber *= 26;
        columnNumber += label.charCodeAt(i) - 'A'.charCodeAt(0) + 1;
    }
    return columnNumber;
}