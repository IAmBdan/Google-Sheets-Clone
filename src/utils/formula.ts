type GetCellValue = (ref: string) => string | number | null;

interface ASTNode {
    type: string;
}

interface LiteralNode extends ASTNode {
    type: 'Literal';
    value: string | number | null;
}

interface FunctionCallNode extends ASTNode {
    type: 'FunctionCall';
    func: string;
    args: ASTNode[];
}

const isNumber = (value: string): boolean => !isNaN(Number(value));
const isString = (value: string): boolean => value.startsWith('"') && value.endsWith('"');
const isReference = (value: string): boolean => /^\$[A-Z]+[0-9]+$/i.test(value);

const evaluate = (expr: string, getCellValue: GetCellValue): string | number | null => {
    if (isReference(expr)) {
        const value = getCellValue(expr);
        if (value === null) throw new Error(`Reference ${expr} not found`);
        return value;
    }
    if (isNumber(expr)) return Number(expr);
    if (isString(expr)) return expr.slice(1, -1).replace(/\\"/g, '"');
    if (expr.startsWith('=')) {
        expr = expr.slice(1);
        return evaluateExpression(expr, getCellValue);
    }
    throw new Error('Invalid expression');
};

const evaluateExpression = (expr: string, getCellValue: GetCellValue): string | number | null => {
    const tokens = tokenize(expr);
    const ast = parseTokens(tokens);
    return evaluateAST(ast, getCellValue);
};

const tokenize = (expr: string): string[] => {
    const regex = /([A-Z]+\(|\)|,|\$[A-Z]+[0-9]+|[0-9.]+|"[^"]*")/gi;
    return expr.match(regex) || [];
};

const parseTokens = (tokens: string[]): ASTNode => {
    let i = 0;

    const parseExpression = (): ASTNode => {
        if (isReference(tokens[i] ?? '') || isNumber(tokens[i] ?? '') || isString(tokens[i] ?? '')) {
            return { type: 'Literal', value: tokens[i++] ?? '' } as LiteralNode;
        }
        if (/^[A-Z]+\($/i.test(tokens[i] ?? '')) {
            const func = (tokens[i++] ?? '').slice(0, -1);
            const args = [];
            while (tokens[i] !== ')' && i < tokens.length) {
                args.push(parseExpression());
                if (tokens[i] === ',') i++;
            }
            if (tokens[i] !== ')') throw new Error('Mismatched parentheses');
            i++;
            return { type: 'FunctionCall', func, args } as FunctionCallNode;
        }
        throw new Error('Unexpected token');
    };

    return parseExpression();
};

const evaluateAST = (node: ASTNode, getCellValue: GetCellValue): string | number | null => {
    switch (node.type) {
        case 'Literal':
            return evaluate((node as LiteralNode).value as string, getCellValue);
        case 'FunctionCall':
            const funcNode = node as FunctionCallNode;
            const args = funcNode.args.map(arg => evaluateAST(arg, getCellValue));
            return evaluateFunctionCall(funcNode.func, args);
        default:
            throw new Error('Unknown AST node type');
    }
};

const evaluateFunctionCall = (func: string, args: (string | number | null)[]): string | number | null => {
    switch (func.toUpperCase()) {
        case 'SUM':
            if (args.some(arg => typeof arg !== 'number')) throw new Error('SUM arguments must be numbers');
            return (args as number[]).reduce((acc, arg) => acc! + (arg as number), 0);
        case 'CONCAT':
            return args.map(arg => (arg !== null ? String(arg) : '')).join('');
        case 'MIN':
            if (args.some(arg => typeof arg !== 'number')) throw new Error('MIN arguments must be numbers');
            return Math.min(...(args as number[]));
        case 'MAX':
            if (args.some(arg => typeof arg !== 'number')) throw new Error('MAX arguments must be numbers');
            return Math.max(...(args as number[]));
        case 'AVG':
            if (args.some(arg => typeof arg !== 'number')) throw new Error('AVG arguments must be numbers');
            return (args as number[]).reduce((acc, arg) => acc! + (arg as number), 0) / args.length;
        case 'IF':
            if (args.length !== 3) throw new Error('IF function requires 3 arguments');
            const condition = args[0];
            const trueValue = args[1] !== undefined ? args[1] : null;
            const falseValue = args[2] !== undefined ? args[2] : null;
            if (condition === undefined || condition === null) throw new Error('IF condition must be defined');
            if (typeof condition !== 'number') throw new Error('IF condition must be a number');
            return condition !== 0 ? trueValue : falseValue;
        default:
            throw new Error(`Unknown function ${func}`);
    }
};

const evaluateFormula = (formula: string, getCellValue: GetCellValue): string | number | null => {
    return evaluate(formula, getCellValue);
};

export { evaluateFormula };
