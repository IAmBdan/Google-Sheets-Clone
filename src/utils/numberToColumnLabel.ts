

export function numberToColumnLabel(number: number): string {
    let label = '';
    while (number > 0) {
        let remainder = (number - 1) % 26;
        label = String.fromCharCode(remainder + 'A'.charCodeAt(0)) + label;
        number = Math.floor((number - 1) / 26);
    }
    return label;
}