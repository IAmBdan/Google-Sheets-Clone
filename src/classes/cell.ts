
export { Cell };
  
import { Ref } from "/Users/bdan/Desktop/Computer Engineering/computer-serve-code/src/classes/ref"

    class Cell {
        value: number | string | { formula: string } | null;
        ref?: Ref;
        refAsString?: string;

        constructor(value: number | string | { formula: string } | null, ref?: Ref | string) {
            this.value = value;
            if (ref) {
                if (typeof ref === 'string') {
                    this.setRefAsString(ref);
                    this.ref = new Ref(ref);
                } else {
                    this.setRef(ref);
                    this.refAsString = ref.toString();
                }
            }
        }


        getValue(): number | string | { formula: string } | null {
            return this.value;
        }

        setValue(value: number | string | { formula: string } | null): void {
            this.value = value;
        }

        setRef(ref: Ref): void {
            this.ref = ref;
            this.refAsString = ref.toString();
        }
        
        setRefAsString(refAsString: string): void {
            this.refAsString = refAsString;
            this.ref = new Ref(refAsString);
        }


        getRef(): Ref {
            if (this.ref) {
                return this.ref;
            } else {
                throw new Error("Ref is undefined");
            }
        }

        getRefString(): string {
            if (this.ref) {
                return this.ref.toString();
            } else {
                throw new Error("Ref is undefined");
            }
        }

        equals(cell: Cell): boolean {
            return this.value?.toString() === cell.value?.toString() && (this.ref?.toString() === cell.ref?.toString()) && this.refAsString === cell.refAsString;
        }
    }