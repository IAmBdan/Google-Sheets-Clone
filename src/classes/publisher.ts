//Brian Daniels
import { hasSpecialChar } from "/Users/bdan/Desktop/Computer Engineering/computer-serve-code/src/utils/hasSpecialChar";

import { hasSpecialChar } from "../utils/hasSpecialChar";
export class Publisher {
  name: string;
  id: number;

  constructor(name: string, id: number) {
    if (name === undefined || hasSpecialChar(name))
      throw new Error("Invalid name");
    this.name = name;
    if (id < 0 || id === -Infinity || id === Infinity || Number.isNaN(id)) {
      throw new Error("Invalid id");
    }
    this.id = id;
  }

  getName(): string {
    return this.name;
  }

  getId(): number {
    return this.id;
  }

  setName(name: string): void {
    if (name === "") {
      throw new Error("Invalid name");
    } else {
      this.name = name;
    }
  }

    setId(id: number): void {
        if(id < 0 || id === -Infinity || id === Infinity || Number.isNaN(id)) {
            throw new Error("Invalid id")
            
        } else {
            this.id = id;
        }
      
    }


  equals(publisher: Publisher): boolean {
    return this.name === publisher.name && this.id === publisher.id;
  }
}
