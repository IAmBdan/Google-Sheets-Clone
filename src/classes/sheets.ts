
import { Cell } from './cell';
import { Ref } from './ref';
import { Term } from "/Users/bdan/Desktop/Computer Engineering/computer-serve-code/types/term"
import { columnToNumber } from '../utils/columnToNumber'
import { set } from 'zod';

 class Sheet {
//should add name and user and shared users list
    private cells: Cell[][];
    private title: string;
    private owner: string;
    private sharedUsers: string[];


    constructor(numColumns: number, numRows: number, title: string, owner: string, sharedUsers: string[]) {
        if (numColumns < 1 || numRows < 1) {
            throw new Error("Invalid sheet dimensions");
        }
        this.cells = Array.from({ length: numRows }, () =>
            Array.from({ length: numColumns }, () => new Cell(null))
        );
        if (title || owner === "") {
            throw new Error("can't be blank");
        }
        this.title = title;
        this.owner = owner;
        this.sharedUsers = sharedUsers;
    }

    getCell(ref: Ref): Cell {
        const columnIndex = ref.getColumnIndex() - 1;
        const rowIndex = ref.row - 1; //assumes rows are 1-indexed, adjust to 0-indexed need to test idk

        if (columnIndex < 0 || columnIndex >= this.cells.length) {
            throw new Error(`Column ${ref.column} is out of bounds`);
        }

        const column = this.cells[columnIndex];
        if (!column || rowIndex < 0 || rowIndex >= column.length) {
            throw new Error(`Row ${ref.row} is out of bounds`);
        }

        const cell = column[rowIndex];
        if (!cell) {
            throw new Error(`Cell at ${ref.row}, ${ref.column} is undefined`);
        }

        return cell;
    }

    getCellByCoords(col: number | string, row: number): Cell {
        let columnIndex: number;

        if (typeof col === "number") {
            columnIndex = col - 1; //adjusts for 1-indexed columns
        } else if (typeof col === "string") {
            columnIndex = columnToNumber(col) - 1; //converts column label to index and adjust for 1-indexed columns, need to test indexes
        } else {
            throw new Error("Invalid column type");
        }

        const rowIndex = row - 1; //adjusts for 1-indexed rows

        if (columnIndex < 0 || columnIndex >= this.cells.length) {
            throw new Error(`Column ${col} is out of bounds`);
        }

        const column = this.cells[columnIndex];
        if (!column || rowIndex < 0 || rowIndex >= column.length) {
            throw new Error(`Row ${row} is out of bounds`);
        }

        const cell = column[rowIndex];
        if (!cell) {
            throw new Error(`Cell at row ${row}, column ${col} is undefined`);
        }

        return cell;
    }

    setCell(ref: Ref, value: Term): void {
        const cell = this.getCell(ref);
        cell.setValue(value);
    }

    getAllUsers(): string[] {
        return [this.owner, ...this.sharedUsers];
    }

    getOwner(): string {
        return this.owner;
    }

    getTitle(): string {
        return this.title;
    }

    getSharedUsers(): string[] {
        return this.sharedUsers;
    }

    setTitle(title: string): void {
        if(title !== "")
        this.title = title;
    }

    setOwner(owner: string): void {
        if(owner !== "")
        this.owner = owner;
    }

    setSharedUsers(sharedUsers: string[]): void {
        this.sharedUsers = sharedUsers;
    }

    addSharedUser(user: string): void {
        if (user !== "" || user !== this.owner || !this.sharedUsers.includes(user))
        this.sharedUsers.push(user);
    }
    addSharedUsers(users: string[]): void {
        this.sharedUsers.push(...users);
    }

}


  