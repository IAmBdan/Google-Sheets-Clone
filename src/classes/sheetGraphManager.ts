import { Publisher } from './publisher'; 
import { Sheet } from './sheets';
// class that will handle the graph of sheets and publishers using ids to locate them
//Brian Daniels

// SheetGraphManager class that manages the relationships between sheets and publishers
export class SheetGraphManager {

    private sheets: Map<number, Sheet>;  // sheet id to sheet
    private publishers: Map<number, Publisher>; // publisher id to publisher
    private sheetPublisherMap: Map<number, number>; // sheet id to publisher id
    private sharedUsersMap: Map<number, number[]>; // publisher id to shared users ids

    /**
     * Sheets (sheets Map):

Key: sheetId (number)
Value: Sheet instance
Publishers (publishers Map):

Key: publisherId (number)
Value: Publisher instance
Sheet-Publisher Relationships (sheetPublisherMap Map):

Key: sheetId (number)
Value: publisherId (number)
Shared Users for Sheets (sharedUsersMap Map):

Key: sheetId (number)
Value: Array of publisherId (number)
     */

    constructor() {
        this.sheets = new Map();
        this.publishers = new Map();
        this.sheetPublisherMap = new Map();
        this.sharedUsersMap = new Map();
    }

    //add a publisher to the map
    addPublisher(publisher: Publisher): void {
        if (this.publishers.has(publisher.getId())) {
            throw new Error("Publisher already exists");
        }
        this.publishers.set(publisher.getId(), publisher);
    }

    //remove a publisher from the map
    removePublisher(publisherId: number): void {
        if (!this.publishers.has(publisherId)) {
            throw new Error("Publisher not found");
        }
        
        for (const [sheetId, pubId] of this.sheetPublisherMap) {
            if (pubId === publisherId) {
                
                this.sheets.delete(sheetId);
                this.sheetPublisherMap.delete(sheetId);
                this.sharedUsersMap.delete(sheetId);
            }
        }
        this.publishers.delete(publisherId);
        
    }

    //add a sheet to the map with a publisher and shared users
    addSheet(sheet: Sheet, publisher: Publisher, sharedUsers: Publisher[] = []): void {
        if (this.sheets.has(sheet.getId())) {
            throw new Error("Sheet already exists");
        }
        if (!this.publishers.has(publisher.getId())) {
            throw new Error("Publisher not found");
        }
    
        this.sheets.set(sheet.getId(), sheet);
        this.sheetPublisherMap.set(sheet.getId(), publisher.getId());
        this.sharedUsersMap.set(sheet.getId(), sharedUsers.map(user => user.getId()));
    }

    //remove a sheet from the map using id
    removeSheet(sheetId: number): void {
        if (!this.sheets.has(sheetId)) {
            throw new Error("Sheet not found");
        }
        //removes the sheet
        this.sheets.delete(sheetId);
        this.sheetPublisherMap.delete(sheetId);

        // Remove the shared users for the sheet, might not need, need to test
        this.sharedUsersMap.delete(sheetId);
    }

    //get a sheet from the map using id
    getSheet(sheetId: number): Sheet | undefined {
        if (!this.sheets.has(sheetId)){
        return undefined; }
        else {
            return this.sheets.get(sheetId);
        }
        
    }

    //get a publisher from the map using id
    getPublisher(publisherId: number): Publisher | undefined {
       if (!this.publishers.has(publisherId)) 
        {return undefined; }
        else  {
            return this.publishers.get(publisherId);
        }   
    }

    //get the publisher for a sheet
    getPublisherForSheet(sheetId: number): Publisher | undefined {
        if (this.sheets.has(sheetId)) {
           
        }

        const publisherId = this.sheetPublisherMap.get(sheetId);
        return publisherId !== undefined ? this.publishers.get(publisherId) : undefined;
    
}

    //get the shared users for a sheet
    getSharedUsersForSheet(sheetId: number): Publisher[] {
        const sharedUserIds = this.sharedUsersMap.get(sheetId);
        return sharedUserIds !== undefined ? sharedUserIds.map(userId => this.publishers.get(userId)!) : [];
    }

    //add a shared user to a sheet
    addSharedUserToSheet(sheetId: number, sharedUser: Publisher): void {
        if (!this.sheets.has(sheetId)) {
            throw new Error("Sheet not found");
        }
        if (!this.publishers.has(sharedUser.getId())) {
            throw new Error("Shared user not found");
        }
        const sharedUsers = this.sharedUsersMap.get(sheetId);
        if (sharedUsers === undefined) {
            throw new Error("Sheet not found");
        }
        sharedUsers.push(sharedUser.getId());
    }

    //remove a shared user from a sheet
    removeSharedUserFromSheet(sheetId: number, sharedUserId: number): void {
        if (!this.sheets.has(sheetId)) {
            throw new Error("Sheet not found");
        }
        const sharedUsers = this.sharedUsersMap.get(sheetId);
        if (sharedUsers === undefined) {
            throw new Error("Sheet not found");
        }
        const index = sharedUsers.indexOf(sharedUserId);
        if (index === -1) {
            throw new Error("Shared user not found");
        }
        sharedUsers.splice(index, 1);
    }

    //get all the sheets for a publisher
    getSheetsForPublisher(publisherId: number): Sheet[] {
        const sheets: Sheet[] = [];
        for (const [sheetId, pubId] of this.sheetPublisherMap) {
            if (pubId === publisherId) {
                sheets.push(this.sheets.get(sheetId)!);
            }
        }
        return sheets;
    }

    //add a sheet to a publisher
    addSheetToPublisher(sheetId: number, publisherId: number): void {
        if (!this.sheets.has(sheetId) || !this.publishers.has(publisherId)) {


        this.sheetPublisherMap.set(sheetId, publisherId);
        //add the shared users to the map
        const sheet = this.getSheet(sheetId);
        const sharedUsers = this.getSharedUsersForSheet(sheetId);
        if (sheet) {
            //add the shared users to the map 
            this.sharedUsersMap.set(sheetId, sharedUsers.map(user => user.getId())); 
            //add the sheet to the publisher
        }
    }
    }

    

    //remove a sheet from a publisher
    removeSheetFromPublisher(sheetId: number): void {
        if (!this.sheets.has(sheetId)) {
            throw new Error("Sheet not found");
        }
        this.sheetPublisherMap.delete(sheetId);
        this.sharedUsersMap.delete(sheetId);
    }




}