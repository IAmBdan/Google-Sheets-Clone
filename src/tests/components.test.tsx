import { expect, describe, it } from "vitest";
import { act, render, screen } from "@testing-library/react";
import Cell from "../app/cell";
import { Ref } from "../classes/ref";
import { Sheet } from "../classes/sheets";
import { Publisher } from "../classes/publisher";
import Grid from "../app/grid";
import { numberToColumnLabel } from "../utils/numberToColumnLabel";
import SheetView from "../app/sheetView";

function makeSheet(content: string | number | { formula: string } | null) {
  const sheet = new Sheet(
    1,
    1,
    "My Sheet",
    new Publisher("Test Publisher", 1),
    [],
  );

  sheet.setCell(new Ref("$A1"), content);

  return sheet;
}

describe("Cell component", () => {
  it("renders correctly", () => {
    const sheet = makeSheet(null);

    const { container } = render(
      <Cell cellRef={new Ref("$A1")} sheet={sheet} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders text", () => {
    const sheet = makeSheet("Hello World!");

    const { container } = render(
      <Cell cellRef={new Ref("$A1")} sheet={sheet} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders numbers", () => {
    const sheet = makeSheet(42);

    const { container } = render(
      <Cell cellRef={new Ref("$A1")} sheet={sheet} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders formulas", () => {
    const sheet = makeSheet({ formula: "=SUM(2,2)" });

    const { container } = render(
      <Cell cellRef={new Ref("$A1")} sheet={sheet} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("updates the value when the cell changes", async () => {
    const sheet = makeSheet(null);

    render(<Cell cellRef={new Ref("$A1")} sheet={sheet} />);

    act(() => {
      sheet.setCell(new Ref("$A1"), "Hello World!");
    });

    expect(screen.getByDisplayValue("Hello World!")).toBeDefined();
  });
});

describe("Grid component", () => {
  it("renders correctly", () => {
    const sheet = makeSheet(null);

    const { container } = render(<Grid sheet={sheet} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the column headers", () => {
    const sheet = new Sheet(
      2,
      100,
      "___",
      new Publisher("Test Publisher", 1),
      [],
    );

    render(<Grid sheet={sheet} />);

    for (let i = 1; i <= 100; i++) {
      expect(screen.getAllByText(numberToColumnLabel(i))).toBeTruthy();
    }
  });

  it("renders the row headers", () => {
    const sheet = new Sheet(
      100,
      2,
      "___",
      new Publisher("Test Publisher", 1),
      [],
    );

    render(<Grid sheet={sheet} />);

    for (let i = 1; i <= 100; i++) {
      expect(screen.getAllByText(i.toString())).toBeTruthy();
    }
  });

  it("renders the appropriate number of cells", () => {
    const sheet = new Sheet(
      13,
      17,
      "___",
      new Publisher("Test Publisher", 1),
      [],
    );

    render(<Grid sheet={sheet} />);

    expect(screen.getAllByRole("textbox")).toHaveLength(13 * 17);
  });
});

describe("SheetView component", () => {
  it("renders correctly", () => {
    const { container } = render(<SheetView />);
    expect(container).toMatchSnapshot();
  });
});
