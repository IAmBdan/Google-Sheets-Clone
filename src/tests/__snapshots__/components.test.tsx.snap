// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cell component > renders correctly 1`] = `
<div>
  <input
    class="border border-black"
    value=""
  />
</div>
`;

exports[`Cell component > renders formulas 1`] = `
<div>
  <input
    class="border border-black"
    value="4"
  />
</div>
`;

exports[`Cell component > renders numbers 1`] = `
<div>
  <input
    class="border border-black"
    value="42"
  />
</div>
`;

exports[`Cell component > renders text 1`] = `
<div>
  <input
    class="border border-black"
    value="Hello World!"
  />
</div>
`;

exports[`Grid component > renders correctly 1`] = `
<div>
  <div
    class="grid h-0 w-full flex-grow overflow-x-scroll overflow-y-scroll"
    style="grid-template-columns: 1fr repeat(1, 1fr); grid-template-rows: 1fr repeat(2, 1fr);"
  >
    <span
      class="border-b border-r border-gray-600 bg-gray-100"
    />
    <span
      class="border-l border-r border-gray-600 bg-gray-100 text-center"
    >
      A
    </span>
    <span
      class="border-b border-t border-gray-600 bg-gray-100 text-center"
    >
      1
    </span>
    <input
      class="border border-black"
      value=""
    />
  </div>
</div>
`;
